### PR TITLE
feat(optimizer): introduce watermark in optimzier

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- add a watermark columns field in plan node's property
- do some derive of some plan nodes.
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/plan_base.rs [hunk 1]
```diff
@@ -72,6 +76,7 @@ impl PlanBase {
         functional_dependency: FunctionalDependencySet,
     ) -> Self {
         let id = ctx.next_plan_node_id();
+        let watermark_cols = FixedBitSet::with_capacity(schema.len());
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/plan_base.rs:79]: Almost all plan nodes set their watermark_cols to the `schema.len()` which means it is a placeholder right now?

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/plan_base.rs:79]: 1. `watermark_cols` only has meaning for the stream plan node. before refactoring the plan node is over, we can not express that type-safely 
2. all the stream plan nodes should try their best to derive the watermark_cols, but giving an empty set can give a correct but not good enough plan.

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/plan_base.rs:79]: Oh, it is an empty set. I get it.

**Review by chenzl25 — APPROVED** [review]
LGTM!

**Pull request merged**